### ISSUE TITLE
Expectation that returns the set of elements without requiring completion

### ISSUE THREAD

**Issue by jchitel** [groue/CombineExpectations#8]
I think there's a gap in the API for when you're dealing with publishers that will never complete. What I'm looking for is a combination between `elements` and `next(_)`: An expectation that will simply return whatever elements have been recorded by the time the timeout completes.

`elements` won't work because it requires the publisher to complete. `next(_)` won't work because it requires you to provide a number of elements to wait for, but if the code under test isn't behaving as expected then there may be more elements that you're missing.

The workaround that I have right now is to use `prefix(_).inverted` with some arbitrary high number like 100. The problem there is (a) I have to provide a number, and (b) the elements aren't consumed. Calling `prefix()` multiple times will include elements that were already returned before.

**Comment by groue** [comment]
Hello @jchitel,

Indeed such an expectation does not exist today. Before we decide if it is a good idea to implement it, could you please provide an actual use case?

**Comment by jchitel** [comment]
An actual use case is SwiftUI's `ObservableObject` with `@Published` properties. For example (this may not actually compile but you get the idea):

```swift
class MyViewModel: ObservableObject {
    @Published var loading: Bool = false
    @Published var data: Int!
    @Published var error: String?

    private var cancellables = Set<AnyCancellable>()

    func load() {
        loading = true
        MyApiService.loadData()
            .sink(
                receiveOutput: { self.data = $0 },
                receiveCompletion: {
                    self.loading = false
                    if let .failure(error) = $0 { self.error = "\(error)" }
                }
            )
            .store(in: cancellables)
    }
}
```

I want to verify the sequence of values set on `loading`, `data`, and `error`, so I can set up recorders on the respective publishers `$loading`, `$data`, and `$error`, which are publishers that will never finish.

The above case is simple, but my actual code is much more complex and incorporates things like pagination and refresh, so I may not be able to predict the exact number of elements to expect from the recorder. This is where it would be nice to have an expectation that just gives me every element that it receives.

**Comment by groue** [comment]
OK, @jchitel. If I were to sum up your request, it would be:

- give me all elements recorded during a given duration, not including elements that were already consumed
- fail only if publisher completes with a failure

**Comment by jchitel** [comment]
Correct

**Comment by madsodgaard** [comment]
+ 1 to adding this feature, would be a really nice addition

**Comment by tonyarnold** [comment]
I could really use this as well. I'm testing bindings to UI code that won't complete, but I'd still like to track the elements for.

**Comment by groue** [comment]
Hello @jchitel, @madsodgaard, @tonyarnold,

Sorry for the late response. I was brought back to this issue by https://twitter.com/tonyarnold/status/1339140819723243521

I'm afraid that the description in https://github.com/groue/CombineExpectations/issues/8#issuecomment-655931907 does not accurately capture your needs, or even that you all have the same needs.

It's not your fault: the available expectations fall in two bags which are not quite easy to distinguish:

- Expectations that cover the whole subscription, from the first published element, independently of other expectations. `prefix` falls in this bag.
- Expectations that build on elements left by previous expectations. `next` is an example.

This makes the task more difficult for you, when you want to express a desire. Maybe the lib will have to lift this ambiguity one day, and make it easier to you to describe your intent. Meanwhile, I have to *guess*.

And my guess is that you all can use an inverted prefix expectation today. It will provide all elements recorded from the beginning of the recording when the timeout expires:

```swift
let recorder = publisher.record()
let elements = try wait(for: recorder.prefix(.max).inverted, timeout: ...)
```

I agree that the intent is not very clear in the above code. If one of you has an idea for a nice name of the `prefix(.max).inverted` expectation, I'm listening.

**Comment by tonyarnold** [comment]
Thanks for the detailed response, @groue! 

> If one of you has an idea for a nice name

Perhaps something like `incompleteElements` or `inFlightElements`? It's similar to the existing `elements` property - although I'm not sure if there needs to be some indication beyond the documentation that this property doesn't consume the elements that are retrieved (assuming that is the behaviour?).

**Comment by groue** [comment]
Hello all, #11 has shipped in v0.6.0: enjoy the new [`availableElements`](https://github.com/groue/CombineExpectations/blob/master/README.md#availableElements) expectation!

**Comment by tonyarnold** [comment]
Thanks, @groue! Happy holidays 😄
